### PR TITLE
array cookie support

### LINKED ISSUE #71: 🔬 Tracking issue for array cookie support
This is a tracking issue for the support of array cookies in TypeART.

### Steps
- [x] Research when and how array cookies are added
- [x] Add test cases for cases where array cookies are expected to be added -> https://github.com/tudasc/TypeART/commit/65684d7912ccbebfce05ac12dae5d587741244ce, https://github.com/tudasc/TypeART/commit/22d8b4f40c018cb89c3e58012525eb77a3fa4d54
- [x] Implementation
  - [x] Finding array cookies -> https://github.com/tudasc/TypeART/commit/d40d567c67e8f8ccaffacb7a1f6a3962057531f5
  - [x] Add array cookie data to existing data structures -> https://github.com/tudasc/TypeART/commit/7781f78e1445e4fc2f2078484dc0269d379833df
  - [x] Consider the padding of array cookies when calculating array element counts -> https://github.com/tudasc/TypeART/commit/0817a3a8296313ca4398d23f6683ce5706f1e2fc
  - [x] In case an array cookie is present, instrument the actual starting address of the array instead of the originally allocated address ->  https://github.com/tudasc/TypeART/pull/73/commits/650b1a395219148fd56ea3aafae1509c96f3e94d
  - [x] ~~Add instrumentation for array cookies~~

### Unresolved Questions
- [x] Should array cookies be instrumented in any way and if so, how? -> for now, only to correctly calculate array sizes

### Array Cookie research
Array cookies are a `size_t` value saving the allocated length of an array. These cookies are allocated under certain conditions when operator new is used to allocate an array.

#### => When are array cookies created?
According to [1], an array cookie is **not** allocated if either of
- "the element type T has a trivial destructor […] and the usual (array) deallocation function […] does not take two arguments" or
- "the `new` operator being used is `::operator new [](size_t, void*)`"

In Clang this is implemented in [`CGCXXABI::requiresArrayCookie`](https://github.com/llvm/llvm-project/blob/llvmorg-10.0.0/clang/lib/CodeGen/CGCXXABI.cpp#L193) using:
- [`CXXNewExpr::doesUsualArrayDeleteWantSize`](https://github.com/llvm/llvm-project/blob/llvmorg-10.0.0/clang/include/clang/AST/ExprCXX.h#L2298) and [`CXXDeleteExpr::doesUsualArrayDeleteWantSize`](https://github.com/llvm/llvm-project/blob/llvmorg-10.0.0/clang/include/clang/AST/ExprCXX.h#L2395) to check whether the delet function takes two arguments.
- [`CXXNewExpr::getAllocatedType`](https://github.com/llvm/llvm-project/blob/llvmorg-10.0.0/clang/include/clang/AST/ExprCXX.h#L2193) as `expr->getAllocatedType().isDestructedType()` to check whether the array element type has a non-trivial destructor.

#### => How are array cookies created?
According to [1]:
- array cookies always have size `sizeof(size_t)`
- if `align` is the maximum alignment of `size_t` and an element of the array and `padding` is the maximum of `sizeof(size_t}` and `align` bytes:
  - "The space allocated for the will be the space required by the array itself plus `padding` bytes"
  - "The cookie will be stored in the `sizeof(size_t)` bytes immediately preceding the array
 In Clang this is implemented in [`ItaniumCXXABI::InitializeArrayCookie`](https://github.com/llvm/llvm-project/blob/llvmorg-10.0.0/clang/lib/CodeGen/ItaniumCXXABI.cpp#L2010).

[1] https://itanium-cxx-abi.github.io/cxx-abi/abi.html#array-cookies

### PR DESCRIPTION
**Pull request by alexanderlinne**
closes #71

### CODE REVIEW

File: lib/passes/analysis/MemOpVisitor.cpp [hunk 1]
```diff
@@ -61,11 +71,22 @@ void MemOpVisitor::visitCallBase(llvm::CallBase& cb) {
...
+  assert(users_it == users.end());
+  assert(isa<T>(first));
+  return dyn_cast<T>(first);
+}
 
-  SmallPtrSet<BitCastInst*, 4> bcasts;
+using MallocGeps   = SmallPtrSet<GetElementPtrInst*, 2>;
+using MallocBcasts = SmallPtrSet<BitCastInst*, 4>;
 
+auto collectRelevantMallocUsers(llvm::CallBase& ci) -> std::pair<MallocGeps, MallocBcasts> {
```
Comment (ahueck) [inline lib/passes/analysis/MemOpVisitor.cpp:87]: Trailing return types (`auto foo() -> bar{..}`) is not common in the code base.
I suggest to keep it as `bar foo() {..}`.

File: lib/passes/analysis/MemOpVisitor.cpp [hunk 2]
```diff
@@ -78,65 +99,85 @@ void MemOpVisitor::visitMallocLike(llvm::CallBase& ci, MemOpKind k) {
...
+
+  assert(array_gep->getNumIndices() == 1);
+  auto array_bcast = expectSingleUser<BitCastInst>(array_gep);
+  bcasts.insert(array_bcast);
+  primary_cast = array_bcast;
+
+  return {ArrayCookieData{cookie_store, array_gep}};
+}
+
+auto handleGepInstrs(MallocGeps const& geps, MallocBcasts& bcasts, BitCastInst*& primary_cast)
```
Comment (ahueck) [inline lib/passes/analysis/MemOpVisitor.cpp:157]: - `T const& foo` is not a style used in TypeART. Suggestion: e.g., `const T& foo`
- `T*& foo` seems unusual. Preference: `T*`

File: lib/passes/instrumentation/MemOpInstrumentation.cpp [hunk 3]
```diff
@@ -74,8 +78,11 @@ InstrCount MemOpInstrumentation::instrumentHeap(const HeapArgList& heap) {
       case MemOpKind::NewLike:
         [[fallthrough]];
       case MemOpKind::MallocLike: {
-        auto bytes   = args.get_value(ArgMap::ID::byte_count);  // can be null (for calloc, realloc)
-        elementCount = single_byte_type ? bytes : IRB.CreateUDiv(bytes, typeSizeConst);
+        elementCount = args.lookup(ArgMap::ID::element_count);
+        if (!elementCount) {
```
Comment (ahueck) [inline lib/passes/instrumentation/MemOpInstrumentation.cpp:82]:
```suggestion
        if (elementCount == nullptr) {
```

**Review by ahueck — COMMENTED** [review]
I have added some minor comments/changes. Overall great patch.

Eventually, we should do a rebase of the branch w.r.t. new changes in devel and then I will squash merge.

**Comment by ahueck** [comment]
Seems like the assert triggers in `expectSingleUser` when compiled with ASAN/UBSAN for the runtime test `41_array_cookie.cpp`.

```
apply.sh 41_array_cookie.cpp 2>&1
opt: /home/ahueck/workspace/typeart/lib/passes/analysis/MemOpVisitor.cpp:79: T *typeart::finder::expectSingleUser(llvm::Value *) [T = llvm::StoreInst]: Assertion `users_it == users.end()' failed.

 #9 0x00007f2c8f9b8747 llvm::StoreInst* typeart::finder::expectSingleUser<llvm::StoreInst>(llvm::Value*) /home/ahueck/workspace/typeart/lib/passes/analysis/MemOpVisitor.cpp:80:3
#10 0x00007f2c8f9b2363 typeart::finder::handleUnpaddedArrayCookie(llvm::SmallPtrSet<llvm::GetElementPtrInst*, 2u> const&, llvm::SmallPtrSet<llvm::BitCastInst*, 4u>&, llvm::BitCastInst*&) /home/ahueck/workspace/typeart/lib/passes/analysis/MemOpVisitor.cpp:124:8
#11 0x00007f2c8f9b28a1 typeart::finder::handleGepInstrs(llvm::SmallPtrSet<llvm::GetElementPtrInst*, 2u> const&, llvm::SmallPtrSet<llvm::BitCastInst*, 4u>&, llvm::BitCastInst*&) /home/ahueck/workspace/typeart/lib/passes/analysis/MemOpVisitor.cpp:160:12

```

**Comment by alexanderlinne** [comment]
@ahueck "get" implies that this function behaves like a normal getter, i.e. I would expect it to not assert. I usually use `assert...` for void function that perform assertions and `expect...` for such functions with a return value.
`LOG_FATAL` should probably just be an assert as well. Otherwise all callees would have to test for nullptr and handle them. Also, the error here, I think, is not that we couln't find a non-asan user, but that we couldn't find a user of the desired type.

File: lib/passes/analysis/MemOpData.h [hunk 4]
```diff
@@ -119,6 +128,7 @@ struct MallocData {
 
 struct FreeData {
   llvm::CallBase* call{nullptr};
+  llvm::GetElementPtrInst* array_cookie_gep{nullptr};  // nullptr iff no array cookie was detected
```
[the remark this answers is not in the document]
Comment (alexanderlinne) [reply lib/passes/analysis/MemOpData.h]: resolved by https://github.com/tudasc/TypeART/pull/73/commits/b143701d6cce961b1ed03836cf2678f9efc7895f

File: lib/passes/analysis/MemOpVisitor.cpp [hunk 5]
```diff
@@ -78,65 +99,85 @@ void MemOpVisitor::visitMallocLike(llvm::CallBase& ci, MemOpKind k) {
...
-    auto dest = bcast->getDestTy();
-    if (!isVoidPtr(dest)) {
-      auto cast = bcast;
-      if (isi64Ptr(dest) && has_specific) {  // LLVM likes to treat mallocs with i64 ptr type, skip if we have sth else
-        continue;
-      }
-      primary_cast = cast;
-    }
+  // We expect bitcasts only after the GEP instructions in this case.
+  assert(bcasts.size() == 0);
```
[the remark this answers is not in the document]
Comment (alexanderlinne) [reply lib/passes/analysis/MemOpVisitor.cpp]: resolved by https://github.com/tudasc/TypeART/pull/73/commits/b143701d6cce961b1ed03836cf2678f9efc7895f

**Pull request merged**